### PR TITLE
Nerv ops

### PR DESCRIPTION
**Pull request by abergeron**
This was done before nervana released their own C interface so maybe we want to switch over to that, but for now, this works.

### CODE REVIEW

**Comment by abergeron** [comment]
This needs https://github.com/Theano/libgpuarray/pull/46 (and probably https://github.com/Theano/Theano/pull/2907 so that the rest works).

**Comment by nouiz** [comment]
travis failed.

**Comment by abergeron** [comment]
It was flake8 and I fixed it.

**Comment by nouiz** [comment]
There need gemm16 tests that are skipped if nervanagpu isn't avail.

**Comment by nouiz** [comment]
```
....
00327	        err = GpuKernel_call(&k_eye_node_8438baf5b42ed491ab3a07bfa420665b_0, 0, 1, 256, args);
00328	        if (err != GA_NO_ERROR) {
00329	            PyErr_Format(PyExc_RuntimeError,
00330	                         "gpuarray error: kEye: %s. n%lu, m=%lu.",
00331	                         GpuKernel_error(&k_eye_node_8438baf5b42ed491ab3a07bfa420665b_0, err),
00332	                         (unsigned long)dims[0], (unsigned long)dims[1]);
00333	            {
00334	        __failure = 7;

....


Problem occurred during compilation with the command line below:
/usr/bin/g++ -shared -g -O3 -fno-math-errno -Wno-unused-label -Wno-unused-variable -Wno-write-strings -DREPLACE_WITH_AMDLIBM -march=core2 -mcx16 -msahf -msse4.1 --param l1-cache-size=32 --param l1-cache-line-size=64 --param l2-cache-size=6144 -mtune=core2 -D NPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -m64 -fPIC -I/u/bastienf/.local/lib64/python2.7/site-packages/pygpu-0.2.1-py2.7-linux-x86_64.egg/pygpu -I/Tmp/lisa/os_v4/anaconda/lib/python2.7/site-packages/numpy/core/include -I/Tmp/lisa/os_v4/anaconda/lib/python2.7/site-packages/numpy/core/include -I/Tmp/lisa/os_v4/anaconda/include/python2.7 -o /tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/8438baf5b42ed491ab3a07bfa420665b.so /tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/mod.cpp -L/Tmp/lisa/os_v4/anaconda/lib -lpython2.7 -lgpuarray -lamdlibm
/tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/mod.cpp: In member function ‘int<unnamed>::__struct_compiled_op_8438baf5b42ed491ab3a07bfa420665b::run()’:
/tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/mod.cpp:327:93: error: invalid conversion from ‘int’ to ‘const size_t*’
/tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/mod.cpp:327:93: error: invalid conversion from ‘int’ to ‘const size_t*’
/tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/mod.cpp:327:93: error: invalid conversion from ‘void**’ to ‘size_t’
/tmp/bastienf/theano.NOBACKUP/compiledir_Linux-2.6-fc14.x86_64-x86_64-with-fedora-14-Laughlin-x86_64-2.7.9-64/tmprO8BjM/mod.cpp:327:93: error: too few arguments to function ‘int GpuKernel_call(GpuKernel*, unsigned int, const size_t*, const size_t*, size_t, void**)’
```

this is with this PR and the new libgpuarray.

**Comment by abergeron** [comment]
Are you "merging" the PR with master?  Because this PR is based from a master that does not have the libgpuarray call changes.

This is a problem of working on multiple separate branches: they won't work by themselves.

**Comment by nouiz** [comment]
I'll try that when you update the PR with my comments.

**Comment by abergeron** [comment]
I'm only missing the Gemm16 tests now.

File: theano/sandbox/gpuarray/nerv.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__)]
+
+    def get_op_params(self):
+        return [('GEMM16_INPLACE', '1' if self.inplace else '0')]
+
+    @staticmethod
+    def cubin_to_code(name):
+        fname = 'hgemm_{0}.cubin'.format(name)
+        with open(os.path.join(nerv.cubin_path, fname)) as f:
```
Comment (nouiz) [inline theano/sandbox/gpuarray/nerv.py]: This fail with my old version of nervanagpu and the new one. The current filename you try to load:

nervanagpu/nervanagpu/kernels/cu/hgemm_nn_128x128.cubin

the current real filename:

nervanagpu/nervanagpu/kernels/cu/hgemm_nn_128x128.cu

**Comment by nouiz** [comment]
There is a problem in debugmode, I'll fix it in another pr.

**Comment by nouiz** [comment]
What diff is needed for the lstm?

**Comment by abergeron** [comment]
This https://github.com/lisa-lab/DeepLearningTutorials/pull/86

**Pull request merged**